File: 0386-lexicographical-numbers/0386-lexicographical-numbers.js
/**
 * @param {number} n
 * @return {number[]}
 */
var lexicalOrder = function(n) {
 class TrieNode {
 constructor(val) {
     this.val = val;
     this.children = {};
     this.isEnd = false;
     this.number = "";
 }
 }

 class Solution {
 constructor() {
     this.root = new TrieNode(".");
 }

 insert(num) {
     let cur = this.root;
     let digits = num.split("");

     for (let digit of digits) {
         if (!(digit in cur.children)) {
             cur.children[digit] = new TrieNode(digit);
         }

         cur = cur.children[digit];
     }

     cur.isEnd = true;
     cur.number = num;
 }

 dfs(curNode, result) {
     if (curNode.isEnd) {
         result.push(curNode.number);
     }

     for (let i = 0; i <= 9; i++) {
         let strI = String(i);
         if (strI in curNode.children) {
             this.dfs(curNode.children[strI], result);
         }
     }
 }

 lexicalOrder(n) {
     for (let num = 1; num <= n; num++) {
         this.insert(String(num));
     }

     let cur = this.root;
     let result = [];
     this.dfs(cur, result);

     let intResult = [];
     for (let num of result) {
         intResult.push(parseInt(num));
     }

     return intResult;
 }
 }

 let solution = new Solution();
 return solution.lexicalOrder(n);

};

File: 0386-lexicographical-numbers/0386-lexicographical-numbers.java
import java.util.*;

class TrieNode {
    char val;
    Map<Character, TrieNode> children;
    boolean isEnd;
    String number;

    public TrieNode(char val) {
        this.val = val;
        this.children = new HashMap<>();
        this.isEnd = false;
        this.number = "";
    }
}

public class Solution {
    private TrieNode root;

    public Solution() {
        this.root = new TrieNode('.');
    }

    public void insert(String num) {
        TrieNode cur = this.root;
        char[] digits = num.toCharArray();

        for (char digit : digits) {
            if (!cur.children.containsKey(digit)) {
                cur.children.put(digit, new TrieNode(digit));
            }

            cur = cur.children.get(digit);
        }

        cur.isEnd = true;
        cur.number = num;
    }

    public void dfs(TrieNode curNode, List<String> result) {
        if (curNode.isEnd) {
            result.add(curNode.number);
        }

        for (int i = 0; i < 10; i++) {
            char n = (char) (i + '0');
            if (curNode.children.containsKey(n)) {
                dfs(curNode.children.get(n), result);
            }
        }
    }

    public List<Integer> lexicalOrder(int n) {
        for (int num = 1; num <= n; num++) {
            insert(String.valueOf(num));
        }

        TrieNode cur = this.root;
        List<String> result = new ArrayList<>();
        dfs(cur, result);

        List<Integer> integerResult = new ArrayList<>();
        for (String num : result) {
            integerResult.add(Integer.parseInt(num));
        }

        return integerResult;
    }
}